tracing/setup.go

```diff
@@ -26,24 +27,38 @@ import (
 	"knative.dev/pkg/tracing/config"
 )
 
-// SetupStaticPublishing sets up trace publishing for the process. Note that other
-// pieces still need to generate the traces, this just ensures that if generated, they are collected
+type Tracer interface {
+	// Shutdown allows for proper cleanup/flush.
+	Shutdown(ctx context.Context) error
+}
+
+// SetupPublishingWithStaticConfig sets up trace publishing for the process. The caller should call .Shutdown() on
+// the returned Tracer before shutdown to make sure all traces are properly flushed.
+// Note that other pieces still need to generate the traces, this just ensures that if generated, they are collected
 // appropriately. This is normally done by using tracing.HTTPSpanMiddleware as a middleware HTTP
 // handler. The configuration will not be dynamically updated.
-func SetupStaticPublishing(logger *zap.SugaredLogger, serviceName string, cfg *config.Config) error {
+func SetupPublishingWithStaticConfig(logger *zap.SugaredLogger, serviceName string, cfg *config.Config) (Tracer, error) {
 	oct := NewOpenCensusTracer(WithExporter(serviceName, logger))
 	if err := oct.ApplyConfig(cfg); err != nil {
-		return fmt.Errorf("unable to set OpenCensusTracing config: %w", err)
+		return oct, fmt.Errorf("unable to set OpenCensusTracing config: %w", err)
 	}
-	return nil
+	return oct, nil
+}
+
+// Deprecated: Use SetupPublishingWithStaticConfig.
+func SetupStaticPublishing(logger *zap.SugaredLogger, serviceName string, cfg *config.Config) error {
+	_, err := SetupPublishingWithStaticConfig(logger, serviceName, cfg)
+	return err
 }
 
-// SetupDynamicPublishing sets up trace publishing for the process, by watching a
-// ConfigMap for the configuration. Note that other pieces still need to generate the traces, this
+// SetupPublishingWithDynamicConfig sets up trace publishing for the process, by watching a
+// ConfigMap for the configuration. The caller should call .Shutdown() on
+// the returned OpenCensusTracer before shutdown to make sure all traces are properly flushed.
+// Note that other pieces still need to generate the traces, this
 // just ensures that if generated, they are collected appropriately. This is normally done by using
 // tracing.HTTPSpanMiddleware as a middleware HTTP handler. The configuration will be dynamically
 // updated when the ConfigMap is updated.
-func SetupDynamicPublishing(logger *zap.SugaredLogger, configMapWatcher configmap.Watcher, serviceName, tracingConfigName string) error {
+func SetupPublishingWithDynamicConfig(logger *zap.SugaredLogger, configMapWatcher configmap.Watcher, serviceName, tracingConfigName string) (Tracer, error) {
```

Comment:
Since we are changing the interface and we want to change all the invocation of `SetupDynamicPublishing` to `SetupPublishingWithDynamicConfig`, could we change to `SetupPublishingWithDynamicConfigAndInitialValue` directly? As I reported in issue https://github.com/knative/pkg/issues/2557, `SetupDynamicPublishing` does not work properly since it doesn't call oct.ApplyConfig(cfg) at the first time.

If you agree, I think we don't need method `SetupPublishingWithDynamicConfig`. We just need method `SetupPublishingWithDynamicConfigAndInitialValue`. 

@dprotaso your thoughts?

Reply:
Since https://github.com/knative/eventing/issues/6456 is not yet implemented and `SetupDynamicPublishingWithInitialValue` is not referenced yet I think it makes sense to do this:
* Keep the current PR as it is and only have the latest/newest `SetupPublishingWithDynamicConfigAndInitialValue`
* If the current PR is merged soon enough the new function can be used directly in a PR for knative/eventing
* If there is a PR for knative/eventing merged before the current one I will put back the backwards-compatible function here so as to not break downstream